plutus-core/plutus-core/src/PlutusCore/Builtin/Elaborate.hs

```diff
@@ -60,40 +60,126 @@ type family GetName k i where
     GetName GHC.Type i = Lookup i '["a", "b", "c", "d", "e", "i", "j", "k", "l"]
     GetName _        i = Lookup i '["f", "g", "h", "m", "n"]  -- For higher-kinded types.
 
--- | Like 'id', but a type constructor.
-type Id :: forall a. a -> a
-data family Id x
+-- | Apply the function stored in the provided 'Maybe' if there's one.
+type MaybeApply :: forall k. Maybe (k -> k) -> k -> k
+type family MaybeApply mayVal x where
+    MaybeApply 'Nothing  a = a
+    MaybeApply ('Just f) a = f a
 
 -- | Try to specialize @a@ as a type representing a PLC type variable.
 -- @i@ is a fresh id and @j@ is a final one (either @i + 1@ or @i@ depending on whether
 -- specialization attempt is successful or not).
--- @f@ is for wrapping 'TyVarRep' (see 'HandleHole' for how this is used).
-type TrySpecializeAsVar :: forall k. Nat -> Nat -> (k -> k) -> k -> GHC.Constraint
-class TrySpecializeAsVar i j f a | i f a -> j
+-- @mw@ is for wrapping 'TyVarRep', if there's a wrapper inside (see 'HandleHole' for how it's
+-- used).
+type TrySpecializeAsVar :: forall k. Nat -> Nat -> Maybe (k -> k) -> k -> GHC.Constraint
+class TrySpecializeAsVar i j mw a | i mw a -> j
 instance
-    ( var ~ f (TyVarRep @k ('TyNameRep (GetName k i) i))
+    ( var ~ MaybeApply mw (TyVarRep @k ('TyNameRep (GetName k i) i))
     -- Try to unify @a@ with a freshly created @var@.
     , a ~?~ var
     -- If @a@ is equal to @var@ then unification was successful and we just used the fresh id and
     -- so we need to bump it up. Otherwise @var@ was discarded and so the fresh id is still fresh.
     -- Replacing @(===)@ with @(==)@ causes errors at use site, for whatever reason.
     , j ~ If (a === var) (i + 1) i
-    ) => TrySpecializeAsVar i j f (a :: k)
+    ) => TrySpecializeAsVar i j mw (a :: k)
+
+type NoAppliedVarsHeader =
+    'Text "A built-in function is not allowed to have applied type variables in its type"
+
+-- | Throw an error telling the user not to apply type variables to anything.
+type ThrowNoAppliedVars :: (GHC.Type -> GHC.Type) -> GHC.Constraint
+type family ThrowNoAppliedVars hole where
+    -- In the Rep context higher-kinded type variables are allowed, but need to be applied via
```

Comment:
Perhaps add a `See Note [Rep vs Type context]` here in case someone finds the "Rep/Type context" terminology unfamiliar.